hack/generator/pkg/astmodel/package_reference_set.go

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import "sort"
+
+// PackageReferenceSet represents a set of distinct PackageReferences
+type PackageReferenceSet struct {
+	references map[PackageReference]struct{}
+}
+
+// NewPackageReferenceSet creates a new empty set of PackageReferences
+func NewPackageReferenceSet(refs ...PackageReference) *PackageReferenceSet {
+	result := &PackageReferenceSet{
+		references: make(map[PackageReference]struct{}),
+	}
+
+	for _, ref := range refs {
+		result.AddReference(ref)
+	}
+
+	return result
+}
+
+// AddReference ensures the set includes an specified Reference
+func (set *PackageReferenceSet) AddReference(ref PackageReference) {
+	set.references[ref] = struct{}{}
+}
+
+// Merge ensures that all references specified in other are included
+func (set *PackageReferenceSet) Merge(other *PackageReferenceSet) {
+	if other == nil {
+		return
+	}
+
+	for ref := range other.references {
+		set.AddReference(ref)
+	}
+}
+
+// Remove ensures the specified item is not present
+// Removing an item not in the set is not an error.
+func (set *PackageReferenceSet) Remove(ref PackageReference) {
+	delete(set.references, ref)
+}
+
+// Contains allows checking to see if an import is included
+func (set *PackageReferenceSet) Contains(ref PackageReference) bool {
+	_, ok := set.references[ref]
+	return ok
+}
+
+// AsSlice() returns a slice containing all the imports
+func (set *PackageReferenceSet) AsSlice() []PackageReference {
+	var result []PackageReference
+
+	if set != nil {
```

Comment:
Is it possible for `set` to be `nil` here? Might be worth leaving a comment if this check is really needed as it reads sorta weird here.

Reply:
It could be **nil**, but I decided that was confusing and removed the nils used elsewhere.